cpp/src/arrow/gpu/cuda_context.cc

```diff
@@ -281,6 +279,49 @@ Result<std::shared_ptr<CudaDevice>> AsCudaDevice(const std::shared_ptr<Device>&
   }
 }
 
+Status CudaDevice::Stream::WaitEvent(const Device::SyncEvent& event) {
+  auto cuda_event =
+      checked_cast<const CudaDevice::SyncEvent*, const Device::SyncEvent*>(&event);
+  if (!cuda_event) {
+    return Status::Invalid("CudaDevice::Stream cannot Wait on non-cuda event");
+  }
+
+  auto cu_event = cuda_event->value();
+  if (!cu_event) {
+    return Status::Invalid("Cuda Stream cannot wait on null event");
+  }
+
+  ContextSaver set_temporary((CUcontext)(context_.get()->handle()));
+  // TODO: do we need to account for CUevent_capture_flags??
```

Comment:
My understanding is the only usage of flags right now is under the context of CUDA Graphs (https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__GRAPH.html#group__CUDA__GRAPH). If they need to handle this case they can always get the raw `Stream` / `Event` handles from Arrow for now and do it themselves. If we need to address it later we can.